crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1717 @@
+            current_shared_file_state: Default::default(),
+        };
+
+        let row_group_decoder = self.init_row_group_decoder();
+        let row_group_decoder = Arc::new(row_group_decoder);
+
+        // Processes row group metadata and spawns I/O tasks to fetch row group data. This is
+        // currently spawned onto the CPU runtime as it does not directly make any async I/O calls,
+        // but instead it potentially performs predicate/slice evaluation on metadata. If we observe
+        // that under heavy CPU load scenarios the I/O throughput drops due to this task not being
+        // scheduled we can change it to be a high priority task.
+        let row_group_data_task_handle = async_executor::spawn(TaskPriority::Low, async move {
+            if start_rx.await.is_err() {
+                drop(row_group_data_fetcher);
+                return metadata_task_handle.await.unwrap();
+            }
+
+            if verbose {
+                eprintln!("[ParquetSource]: Starting row group data fetch")
+            }
+
+            // We must `recv()` from the `NotifyReceiver` before awaiting on the
+            // `normalized_slice_oneshot_rx`, as in the negative offset case the slice resolution
+            // only runs after the first notify.
+            if !row_group_data_fetcher.init_next_file_state().await {
+                drop(row_group_data_fetcher);
+                return metadata_task_handle.await.unwrap();
+            };
+
+            let slice_range = {
+                let Ok(slice) = normalized_slice_oneshot_rx.await else {
+                    // If we are here then the producer probably errored.
+                    drop(row_group_data_fetcher);
+                    return metadata_task_handle.await.unwrap();
+                };
+
+                slice.map(|(offset, len)| offset..offset + len)
+            };
+
+            row_group_data_fetcher.slice_range = slice_range;
+
+            // We can use buffer_unordered as the row group data contains the morsel seq.
+            let mut row_group_stream = row_group_data_fetcher
+                .into_stream()
+                .map(|x| async {
+                    match x {
+                        Ok(handle) => handle.await,
+                        Err(e) => Err(e),
+                    }
+                })
+                .buffer_unordered(row_group_prefetch_size)
+                .map(|x| async {
+                    let row_group_decoder = row_group_decoder.clone();
+                    match x {
+                        Ok(row_group_data) => {
+                            async_executor::spawn(TaskPriority::Low, async move {
+                                row_group_decoder.row_group_data_to_df(row_group_data).await
+                            })
+                            .await
+                        },
+                        Err(e) => Err(e),
+                    }
+                })
+                .buffer_unordered(num_pipelines);
```

Comment:
@orlp  This changes the decode back to spawning tasks onto the executor to support splitting a single row group into multiple morsels. This lets me distribute the morsels more evenly across the pipelines - with the previous approach, I think if I added the splitting in-place, then even after I split the row group into several morsels they would still end up being sent serially across the same pipeline.

Reply:
> I think if I added the splitting in-place, then even after I split the row group into several morsels they would still end up being sent serially across the same pipeline.

I don't understand why, as long as the splitting happens before going into the work distributor, everything should be fine. Is the splitting itself also computationally intensive?

Reply:
It's because we can't split the raw byte-data of the row group, so instead I'm splitting the row group after it's been decoded into a DataFrame. In the previous version the row group decoding took place after the work distributor was used to distribute the raw row group byte data - at that point I was no longer able to re-distribute the individual splits within a row group.

Reply:
Hmm.. I thought about it again and I don't like what I'm currently doing either.. it's not receiving backpressure properly from the pipeline.

I think it's better to leave it as 1 morsel per row-group for now?